### PR TITLE
Backport 2287 to release/0.4.z

### PR DESCRIPTION
**Pull request by mrrajan**
Manual backport of PR #2287 to the release/0.4.z branch.

 The automated backport failed due to merge conflicts. This PR contains the manually cherry-picked and conflict-resolved commits.

 ## Changes

 This backport includes 14 commits from PR #2287 and PR #2242

 ## Resolved Conflicts
 Merge conflicts were resolved in:
 - Entity and migration module imports (sbom_group, sbom_group_assignment, sbom_license_expanded)
 - License query service refactoring (union queries for SPDX and non-SBOM licenses)
 - Database schema updates (MigratorExt trait, migration list)
 - SBOM service imports (sbom_license_expanded, expanded_license)
 - Recursion limit attributes for CI coverage builds
 - Import cleanup in cyclonedx.rs

 Custom Fix:
 - IngestResult.id is Id enum on release/0.4.z, not String. Changed Uuid::parse_str(&result.id) to result.id.try_as_uid() in modules/fundamental/src/license/service/test.rs, modules/fundamental/src/sbom/service/test.rs
 - Commit 49185eb8 rewrote licenses() but lost NOT EXISTS optimization from 3e9348d5. Query reverted to inefficient LEFT JOIN + IS NULL (scans all rows). Applied NOT EXISTS pattern. Changed from sbom_package_license to sbom_license_expanded table on modules/fundamental/src/license/service/mod.rs

Summary                                                                                                                                                                                                          
                  
  Fixes compatibility issues and restores performance optimizations lost during the backport of PR #2287 to release/0.4.z.                                                                                         


## Summary by Sourcery

Normalize and pre-expand license expressions into dedicated tables, refactor license querying and filtering across services to use the new schema and COALESCE-based expressions, and add supporting migrations, entities, and tests while increasing recursion limits for large builds.

New Features:
- Introduce expanded_license and sbom_license_expanded tables and entities to store normalized expanded license texts and their SBOM/license mappings.
- Populate expanded license data at SBOM ingestion time for both SPDX and CycloneDX documents using a dedicated ingestion helper.
- Add comprehensive tests validating license union queries, COALESCE behavior, junction-table integrity, and license-based filtering across SBOM and PURL services.

Bug Fixes:
- Fix inconsistent behavior between SPDX and CycloneDX license handling by unifying them on a normalized, pre-expanded license representation.
- Eliminate duplicate expanded license entries by deduplicating on an MD5 hash of the expanded text.

Enhancements:
- Refactor license listing, SBOM, and PURL services to use pre-expanded license data via COALESCE expressions instead of on-the-fly expansion functions and complex CTEs.
- Simplify and optimize license-based filtering for SBOMs, packages, and PURLs by replacing custom CASE/CTE logic with unioned subqueries against normalized license data.
- Deprecate the licenses_ref_mapping field in SBOM models and OpenAPI schemas now that licenses are pre-expanded and the mapping is always empty.
- Extend the migration framework with a MigratorExt abstraction and register new license-related migrations.
- Increase recursion limits across several crates and test modules to support deeper macro expansions and large code paths.

Tests:
- Add new integration tests for license services and SBOM services covering COALESCE expansion, filtering, pagination, ordering, and ingestion-time population of expanded licenses.
- Add component type parsing/serialization tests for CycloneDX component types.

## Summary by Sourcery

Backport license normalization and querying improvements to the 0.4.z release branch, including new expanded license storage, ingestion-time expansion, and performance-optimized license filtering across SBOM and PURL services.

New Features:
- Introduce expanded_license and sbom_license_expanded tables and entities to store normalized expanded license texts and their SBOM/license mappings.
- Populate expanded license data during SPDX and CycloneDX SBOM ingestion using a dedicated helper.

Enhancements:
- Refactor license listing and license-based filtering in license, SBOM, and PURL services to use pre-expanded license data with COALESCE expressions and NOT EXISTS optimizations.
- Deprecate licenses_ref_mapping in SBOM models and OpenAPI schemas now that licenses are pre-expanded and mappings are always empty.
- Add database indexes and migrations to improve license query performance and normalize expanded license storage.
- Increase recursion limits for several crates to support larger builds and deeper macro expansions.

Tests:
- Add extensive tests for license union queries, COALESCE behavior, junction-table integrity, pagination/counting, and license-based filtering across services.
- Add SBOM and PURL service tests validating COALESCE-based license aggregation and filtering, and coverage for edge cases like overlapping LicenseRef identifiers and NOT NULL filters.

### CODE REVIEW

**Review by ctron — CHANGES_REQUESTED** [review]
That seems to pull in all kinds of stuff from main which should not be in the PR.

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,10 @@
 
 *.cdx.json
 *.folded
+*-secret.*
+*.tar
+*.sql
```
Comment (ctron) [inline .gitignore]: I'd not want to exclude all `.sql`, we might have `.sql` files elsewhere. Maybe not now. But maybe in the future. (like `./modules/fundamental/src/sbom/service/gc_purls_after_sbom_deletion.sql`).

**Review by ctron — COMMENTED** [review]
Minor nit with the ignores.

However, I'm just saying that that PR has quite an impact in how we handle query as well as ingestion. And we don't have any performance measurements, as far as I know.

And the goal here is to merge this into a patch release.

I think it would make sense to, at least manually, give it a try and see how it impacts performance with a larger set of data.

File: .gitignore [hunk 2]
```diff
@@ -11,7 +11,7 @@
 *.folded
 *-secret.*
 *.tar
-*.sql
+#*.sql
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply .gitignore]: Removed the lines.

**Review by ctron — COMMENTED** [review]
Thanks ❤️  Technically that PR is fine. I think someone needs to make a decision if we want to this merged without performance checks. Unless we already have some somewhere.

**Comment by mrrajan** [comment]
Closing the PR due to conflicts and scope of the fix defined for release/0.5.z.
